### ISSUE TITLE
audio play() called before user interaction

### ISSUE THREAD

**Issue by AnmAtAnm** [google/blockly#1902]
New requirements in Chrome 67 require the user to interact with the page before Audio.play() will succeed.  The call to `play()` in `Blockly.WorkspaceAudio.preload()` now fails with the following message:

    DOMException: play() failed because the user didn't interact with the document first.

**Comment by NeilFraser** [comment]
There is code already present which is supposed to handle that.
inject.js lines 403-409

I guess 'mousemove' is no longer considered an interaction.

**Comment by AnmAtAnm** [comment]
> I guess 'mousemove' is no longer considered an interaction.
Correct.  Now it requires a click.

**Comment by ewpatton** [comment]
We often embed assets (e.g., the cursors) as data URIs so that they are available immediately after the sources are loaded (although not in this particular case). Would that work for Blockly's use case? One concern might be bloating the compiled version, but you could have a separate JS file with just the base64 encoded assets.

**Comment by AnmAtAnm** [comment]
@ewpatton Interesting idea.  We currently attempt to minimize the audio download size by offering multiple audio file formats (each platform has their own list of compatible of file format and compression formats). It probably wouldn't make sense to do that if we were embedding the files in the code, especially with the +33% increase that base64 encoding will likely incur.

If someone could do an updated evaluation of supported audio formats, going back to IE 11 and including mobile platforms, we could understand if we could reduce the number of required audio files (ideally down to three, one per audible event). (Note to self: make a test page of all nine audio files as data URIs.)

If this is true, we could use your proposed method and get audio on offline instances without requiring the app developer to define the `media/` directory option.

**Comment by ewpatton** [comment]
@AnmAtAnm I would expect that the supported audio types will be fixed for a given (os, browser, version) tuple. One could have a JS file per audio type and detect which script to load at runtime. Here's a rough sketch of what I am thinking:

```
// sounds.mp3.js:
Blockly.SoundData["click"] = "data:audio/mp3;base64,...";
...
```

```
// sounds.ogg.js:
Blockly.SoundData["click"] = "data:audio/ogg;base64,...";
...
```

```
// sounds.wav.js:
Blockly.SoundData["click"] = "data:audio/wav;base64,...";
```

```
// workspace_audio.js:
Blockly.WorkspaceAudio.loadSoundData = function(preferredFormats) {
  for (var i = 0; i < preferredFormats.length; i++) {
    if (/* test for preferredFormat[i] support */) {
      // Add <script> tag for formatted sound data.
    }
  }
}
```

**Comment by rachel-fenichel** [comment]
Figuring out whether a given audio type is playable is pretty inexact.  You use `canPlayType`, and it returns a string. Possible values are:

`'probably'`: The specified media type appears to be playable.
`'maybe'`: Cannot tell if the media type is playable without playing it.
`''` (empty string): The specified media type definitely cannot be played.

MDN page [here](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/canPlayType).

Using a dataurl does look like a good way to move this check earlier, but it looks like you need at least one actual media object to check if it works (in the case of `maybe`).

**Comment by ewpatton** [comment]
I hacked together sample audio.js files based on my proposal above just to provide some numbers:

| format | media (raw) | media (gzip) | js (raw) | js (gzip) |
|---|---|---|---|---|
| mp3 | 6.8k | 4.4k | 9.3k | 4.3k |
| ogg | 14.7k | 12.2k | 18k | 11k |
| wav | 14.1k | 11.6k | 17k | 12k |

It seems like with GZIP compression on the connection there might even be a slight file size savings with this technique, at least in the mp3 and ogg cases. To address @rachel-fenichel's comment, there could be a file with dummy sounds (say 10-20ms) that could be used for testing codec support that should also be relatively lightweight. Alternatively, with long cache times (none of the audio files have changed since they were added), one could load all three formats and play whichever works first.

**Comment by f0zi** [comment]
I think the problem is not the auto-play, it's the unhandled promise rejection. Adding `.catch(()=>{})` to the `play()` call in `Blockly.WorkspaceAudio.prototype.preload` fixes the error, and I think it's a valid solution because:

- It's effectively what you are doing already: You are ignoring the result. So it's not worse than before.
- I'm not sure about this, but I think even though the sound did not play it got preloaded.
- It's actually what [Google recommends](https://developers.google.com/web/updates/2017/09/autoplay-policy-changes#best-practices). Just go ahead and do it, catch the error, and *then* act on it.

**Comment by samelhusseini** [comment]
Fixed by #2162

**Comment by quantuminformation** [comment]
What if the user clicks anywhere on the page, can you then play the sound later any time you want with eg

     var audio = new Audio("onmessage.mp3")
            audio.play()
